wordpress-mobile/WordPress-iOS
# Spotlight: Index posts and pages

This PR allows posts and pages to be indexed with core spotlight:

![img_092e4a428628-1](https://user-images.githubusercontent.com/154014/37435966-aa23e94c-27b3-11e8-9c25-3988d05226fb.jpeg)![img_06981d7d05a1-1](https://user-images.githubusercontent.com/154014/37435975-b532512a-27b3-11e8-9d2e-ab789268cf04.jpeg)

[Demo Video is here](https://cloudup.com/c7zIMcC8eeu).

## Discussion
A new class has been created to manage all aspects of indexing: `SearchManager.swift`. Anything that can be indexed should adopt the `SearchableItemConvertable` protocol. In this PR, `AbstractPost` is the only item that adopts it (via `AbstractPost+Searchable.swift`).

Fixes #8781 
Ref: #8786 

### UI
The spotlight items follow the [standard UI that iOS defines](https://developer.apple.com/library/content/documentation/General/Conceptual/AppSearch/Art/numbers_2x.png). We alter the description text based on the data we have available to us:

<img width="844" alt="napkin 63 03-14-18 6 32 44 pm" src="https://user-images.githubusercontent.com/154014/37436476-1ad013ee-27b6-11e8-916d-adf7ddb23c13.png">

### Indexing
After some discussion with @iamthomasbishop, we decided to index Posts and Pages as a user sees them in the Post/Page list VCs. Basically, as batches of `AbstractPost`s are synced down from the server, we push them into the index. This will ensure that spotlight contains content the user actually saw in the app making the search more relevant and timely. We index all pages/posts for both .org and .com sites.

**Note 1:** A post/page with any status except `trash` will be indexed.
**Note 2:** This is a private, on-device index only and is not synced via iCloud between devices.

### Pruning
* If a post/page is trashed, it is removed from the spotlight index (and if it is restored it is re-added back into the index)
* If a site is hidden or removed, all of the content associated with that site is removed from the index
* If a user logs out, the entire spotlight index is cleared (for WPiOS)
* If a user taps on the "Clear Spotlight Index" button is App settings, the entire spotlight index is cleared (for WPiOS).

### Updating
If a post/page is created or updated, it's content will be updated in spotlight.

### Selecting a spotlight result
If a post or page result is tapped in spotlight, WPiOS will navigate to the appropriate Post/Page list and then open the editor with the selected Post/Page.

### Analytics
We are tracking 2 new metrics: `WPAnalyticsStatSpotlightSearchOpenedPost` and `WPAnalyticsStatSpotlightSearchOpenedPage` which are logged after a user launches an app through a spotlight search.

## Testing
Testing should be done on a real device (no sim) — spotlight is too unpredictable on the simulator to be an accurate guage for testing.

Generic testing steps:
1) Go to Site Details-->Pages and/or Posts)
2) Scroll though the list and allow items to be indexed.
3) Open spotlight search in iOS and start looking for posts/pages from the site you were just on.
4) Tap on a spotlight result and make sure it opens the correct page/post.

Other testing ideas:
* Index a site's post and then hide the site in WPiOS — the posts/pages should no longer be in the index
* Be sure to test .org and .com and Jetpack sites. Make sure all of the post/page context is indexed.
* Try going to App Settings and pressing the "Clear Spotlight Index" button. Make sure it works (and it will also help you in testing)
* Index a post, find it in spotlight, and then update it with new content. Verify the post is updated in spotlight.
* Create a new post/page — verify it is added to the spotlight index.
* Trash a post, verify it is removed from spotlight.
* Restore a trashed post, verify it is added into spotlight.

## Review
@etoledom would you mind taking a peek 👀 at this for me? Thanks!
/cc @frosty, @kwonye and @iamthomasbishop for design
